torch/nn/functional.py

```diff
@@ -3953,3 +3953,188 @@ def multi_head_attention_forward(query,                           # type: Tensor
+                training=training, key_padding_mask=key_padding_mask, attn_mask=attn_mask)
+    batch_heads, tgt_len, head_dim = q.size()
+    assert q.size(0) == k.size(0) == v.size(0), "Dimension 0 of q, k, v must be equal."
+    assert batch_heads % num_heads == 0, "Dimension 0 of q, k, v must be divisible by num_heads"
+    bsz = batch_heads // num_heads
+    assert k.size() == v.size(), "Shape of k, v must match"
+    assert q.size(-1) == k.size(-1), "The head dimension of query must be equal to that of key"
+
+    src_len = k.size(1)
+
+    # Scale q
+    q = q * (float(head_dim) ** -0.5)
+    if attn_mask is not None:
+        if attn_mask.dim() == 2:
+            attn_mask = attn_mask.unsqueeze(0)
+            if list(attn_mask.size()) != [1, tgt_len, src_len]:
+                raise RuntimeError('The size of the 2D attn_mask is not correct.')
+        elif attn_mask.dim() == 3:
+            if list(attn_mask.size()) != [batch_heads, tgt_len, src_len]:
+                raise RuntimeError('The size of the 3D attn_mask is not correct.')
+        else:
+            raise RuntimeError("attn_mask's dimension {} is not supported".format(attn_mask.dim()))
+        # attn_mask's dim is 3 now.
+        if attn_mask.dtype == torch.bool:
+            attn_mask = torch.where(
+                attn_mask, torch.tensor(float('-inf')), torch.tensor(0.)).to(dtype=q.dtype, device=q.device)
+
+    src_len = k.size(1)
+
+    if key_padding_mask is not None:
+        assert key_padding_mask.size(0) == bsz
+        assert key_padding_mask.size(1) == src_len
+
+    if add_zero_attn:
+        src_len += 1
+        k = torch.cat([k, torch.zeros((k.size(0), 1) + k.size()[2:], dtype=k.dtype, device=k.device)], dim=1)
+        v = torch.cat([v, torch.zeros((v.size(0), 1) + v.size()[2:], dtype=v.dtype, device=v.device)], dim=1)
+        if attn_mask is not None:
+            attn_mask = pad(attn_mask, (0, 1))
+        if key_padding_mask is not None:
+            key_padding_mask = pad(key_padding_mask, (0, 1))
+
+    # Dot product of q, k
+    attn_output_weights = torch.matmul(q, k.transpose(-2, -1))
+    assert list(attn_output_weights.size()) == [batch_heads, tgt_len, src_len]
+
+    if attn_mask is not None:
+        attn_output_weights += attn_mask
+
+    if key_padding_mask is not None:
+        attn_output_weights = attn_output_weights.view(bsz, num_heads, tgt_len, src_len)
+        attn_output_weights = attn_output_weights.masked_fill(
+            key_padding_mask.unsqueeze(1).unsqueeze(2),
+            float('-inf'),
+        )
+        attn_output_weights = attn_output_weights.reshape(batch_heads, tgt_len, src_len)
+
+    attn_output_weights = softmax(attn_output_weights, dim=-1)
+
+    attn_output = torch.matmul(dropout(attn_output_weights, p=dropout_p, training=training), v)
+    return attn_output, attn_output_weights
+
+
+def multi_head_attention_out_projection(attn_output, num_heads, out_proj_weight, out_proj_bias=None):
```

Comment:
Are all of these functions scriptable?

Reply:
We should have the JIT support for all the three new functions.